### DISCUSSION TITLE
Page Buffer 128x128 HW i2c loop time over 1.5 seconds normal?

### DISCUSSION THREAD

**Discussion by qfleuren** [olikraus/u8g2#1516] [Q&A]
I'm wondering if it's normal for a screen with 128x128 pixels to take about 1400ms to write? 
Or am I doing stupid things with my code making the drawing process take longer than it should? 
This is my first time using this lib and screen for an actual project, tests were quick, but I'm guessing those didn't have this many commands? :D 

Any help or pointers appreciated!

This is the constructor I've used with buffer mode 2.
`U8G2_SSD1327_MIDAS_128X128_2_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE); // HW I2C pagebuffer`

this is my ''setup''  for the screen 
`void screensetup() {
  u8g2.begin();
}`
``` C++
void sensor_select_screen(int sensor) {
  u8g2.firstPage();
  do {
    u8g2.drawLine(1, 26, 128, 26);
    u8g2.drawLine(100, 1, 100, 25);
    // pijl
    u8g2.drawBox(107, 7, 10, 10);
    u8g2.drawTriangle(116, 3, 116, 21, 127, 12);
    // text
    u8g2.setFont(u8g2_font_5x8_tf);
    u8g2.drawStr(2, 15, sensortitel[sensor]);
    u8g2.setFont(u8g2_font_7x14_tf);
    u8g2.drawStr(25, 50, "Select sensor");
  }
  while ( u8g2.nextPage() );
}
```

And this is the function of my program to actually write my sensory data onto my screen depending on my program situation.

``` C++
void screen_write (int sensor) {
  u8g2.setFont(u8g2_font_ncenB14_tr);
  u8g2.firstPage();
  do {
    u8g2.setFontMode(1);  /* activate transparent font mode */
    u8g2.setDrawColor(1); /* color 1 for the box */
    // horizontale lijnen
    u8g2.drawLine(1, 26, 128, 26);
    u8g2.drawLine(1, 110, 128, 110);
    // verticale lijnen x,y x,y
    u8g2.drawLine(100, 1, 100, 25);
    // I/o lijnen
    u8g2.drawBox(0, 110, 17, 18);
    u8g2.drawLine(111, 110, 111, 128);
    for ( int i = 16; i < 112; i = i + 16) {
      u8g2.drawLine(i, 110, i, 128);
    }
    // pijl
    u8g2.drawBox(107, 7, 10, 10);
    u8g2.drawTriangle(116, 3, 116, 21, 127, 12);
    // text
    u8g2.setFont(u8g2_font_5x8_tf);
    u8g2.drawStr(114, 124, "GND");
    u8g2.drawStr(32, 36, "Sensor Data:");
    u8g2.setDrawColor(0); /* color 1 for the box */
    u8g2.drawStr(1, 124, "I/O");
    u8g2.setDrawColor(1); /* color 1 for the box */
    u8g2.drawStr(2, 15, sensortitel[sensor]);
    //DATA TEXT
    u8g2.setFont(u8g2_font_micro_mr);
// Sensor I/o weg schrijven naar OLED
    for (int x = 0; x < 6; x++) {
      u8g2.drawStr((21 + ((16 * x) - 2)), 123, sensorio[sensorselect - 1][x]);
      }
// Sensor data+units  wegschrijven naar OLED
    for (int x = 0; x < sensordatatypes[sensorselect]; x++) {
      switch (x) {
        case 0:
 u8g2.drawStr(5,50, sensorunits[sensorselect-1][x]);
 if (sensorselect ==3) { // loadcell data convert
  u8g2.drawStr(40,50, String((sensordata[x]/10.0)).c_str() );
 }
 else if (sensorselect ==5) { // temp sensor data convert
  u8g2.drawStr(40,50, String((sensordata[x]/10.0-40)).c_str() );
 }
 else 
 {
 u8g2.drawStr(40,50, String(sensordata[x]).c_str() );
 }
          break;
        case 1:
u8g2.drawStr(69,50, sensorunits[sensorselect-1][x]);
u8g2.drawStr(99,50, String(sensordata[x]).c_str() );
          break;
        case 2:
u8g2.drawStr(5,90, sensorunits[sensorselect-1][x]);
u8g2.drawStr(40,90, String(sensordata[x]).c_str() );
          break;
        case 3:
u8g2.drawStr(69,90, sensorunits[sensorselect-1][x]);
u8g2.drawStr(99,90, String(sensordata[x]).c_str() );
          break;
        default:
          break;
      }
    }
  }
    while ( u8g2.nextPage() );
  }
```

**Comment by olikraus** [comment]
- The OLED itself is slow. It is a 4-bit graylevel OLED, so for each pixel, 4 bit have to be transfered (so transfer is 4 times slower compared to a SSD1306 OLED)
- Your OLED has 128x128 pixel. This is twice as much compared to a SSD1306 128x64 OLED. Together with the first point this makes the transfer 8x slower
- I2C is slower than SPI. Maybe you can improve this a little bit with setBusClock command, but there are some physical limitations with I2C compared to SPI which makes SPI (much?) slower than SPI
- Try to compute as much as possible outside the picture loop: For example `String((sensordata[x]/10.0)).c_str()` is slow. Try to precalculate this and only use the resulting value within the u8g2 while loop.

Update: The SSD1327 seems to be very common, however for a 128x128 display the SH1107 seems to be a much faster solution, so maybe look out for a 128x128 OLED with SH1107 controller.

**Reply by qfleuren** [reply]
First of all thanks for responding to my question!
I was just unsure because the loops were taking 4.5 seconds originally with SW i2c 
but setting buffer 2+ HW i2c + setBusClock(30000) and moving one of the font functions out of the do-while loop lowered the
delay to around 420ms which is totally acceptable for most use cases! 

for the current project, however, I've rewritten the needed commands to an if-else structure so the main loop and all its sensors take prio over screenwriting. this takes the delay down to 60ms which is more than enough for what I need it for.^^
